File: src/retriever.py
"""
High-level search interface. Wraps SpecStore with convenient filtering helpers
and is the single import used by both the CLI and the MCP server.
"""

from __future__ import annotations

from typing import Optional

from .store import SpecStore


class Retriever:
    def __init__(self, store: Optional[SpecStore] = None):
        self._store = store or SpecStore()

    # ------------------------------------------------------------------
    # Ingest
    # ------------------------------------------------------------------

    def ingest(self, source: str, api_name: str) -> dict:
        """Load an OpenAPI spec and push all endpoints + schemas to ChromaDB.

        Returns a summary dict: { api, endpoints_ingested, schemas_ingested, total }.
        """
        from .parser import load_spec, extract_endpoints, extract_schemas
        from .chunker import endpoint_to_document, schema_to_document

        spec = load_spec(source)
        endpoints = extract_endpoints(spec)
        schemas = extract_schemas(spec)

        endpoint_docs = [endpoint_to_document(e, api_name) for e in endpoints]
        schema_docs = [schema_to_document(s, api_name) for s in schemas]

        self._store.delete_api(api_name)
        self._store.upsert(endpoint_docs + schema_docs)

        return {
            "api": api_name,
            "endpoints_ingested": len(endpoint_docs),
            "schemas_ingested": len(schema_docs),
            "total": len(endpoint_docs) + len(schema_docs),
        }

    # ------------------------------------------------------------------
    # Search
    # ------------------------------------------------------------------

    def search_endpoints(
        self,
        query: str,
        api: Optional[str] = None,
        method: Optional[str] = None,
        tag: Optional[str] = None,
        n: int = 5,
    ) -> list[dict]:
        """Semantic search over endpoints with optional metadata filters."""
        where = _build_where(type_="endpoint", api=api, method=method)
        results = self._store.query(query, n_results=n, where=where or None)
        # Post-filter by tag (ChromaDB doesn't support substring matching natively)
        if tag:
            tag_lower = tag.lower()
            results = [
                r for r in results
                if tag_lower in r["metadata"].get("tags", "").lower()
            ]
        return results

    def search_schemas(
        self,
        query: str,
        api: Optional[str] = None,
        n: int = 5,
    ) -> list[dict]:
        """Semantic search over schemas with optional API filter."""
        where = _build_where(type_="schema", api=api)
        return self._store.query(query, n_results=n, where=where or None)

    def get_endpoint(
        self,
        path: str,
        method: str,
        api: Optional[str] = None,
    ) -> Optional[dict]:
        """Exact lookup of a specific endpoint by path + method (+ optional api)."""
        if api:
            doc_id = f"{api}:endpoint:{method.upper()}:{path}"
            return self._store.get_by_id(doc_id)
        # Search across all apis
        where = {"$and": [{"type": "endpoint"}, {"method": method.upper()}, {"path": path}]}
        results = self._store.query(f"{method.upper()} {path}", n_results=1, where=where)
        return results[0] if results else None

    # ------------------------------------------------------------------
    # Metadata
    # ------------------------------------------------------------------

    def delete_api(self, api_name: str) -> None:
        self._store.delete_api(api_name)

    def list_apis(self) -> list[str]:
        return self._store.list_apis()

    def list_endpoints(self, api_name: str) -> list[dict]:
        """Return all endpoint documents for a given API."""
        docs = self._store.get_all(api_name)
        return [d for d in docs if d.get("metadata", {}).get("type") == "endpoint"]


# ---------------------------------------------------------------------------
# Internal helpers
# ---------------------------------------------------------------------------

def _build_where(**filters) -> dict:
    """Build a ChromaDB `where` clause from keyword filters, omitting None values."""
    clauses = []
    type_ = filters.get("type_")
    if type_:
        clauses.append({"type": type_})
    api = filters.get("api")
    if api:
        clauses.append({"api": api})
    method = filters.get("method")
    if method:
        clauses.append({"method": method.upper()})

    if not clauses:
        return {}
    if len(clauses) == 1:
        return clauses[0]
    return {"$and": clauses}

File: packages/mcp-server/src/core/retriever.ts
import SpecStore from "./store";
import { loadSpec, parseSpecContent, extractEndpoints, extractSchemas } from "./parser";
import { endpointToDocument, schemaToDocument } from "./chunker";
import type { ApiInfo, DocumentResult, IngestSummary, QueryResult } from "#types/store";

export interface ProgressEvent {
	phase: "parsing" | "parsed" | "deleting" | "embedding" | "storing" | "done";
	message: string;
	done?: number;
	total?: number;
}

// ---------------------------------------------------------------------------
// Retriever
// ---------------------------------------------------------------------------

export default class Retriever {
	#store: SpecStore;

	constructor(store?: SpecStore) {
		this.#store = store ?? new SpecStore();
	}

	// ------------------------------------------------------------------
	// Ingest
	// ------------------------------------------------------------------

	async ingest(
		source: string,
		apiName: string,
		onProgress?: (event: ProgressEvent) => void,
		opts?: { skipDelete?: boolean },
	): Promise<IngestSummary> {
		onProgress?.({ phase: "parsing", message: "Loading spec..." });
		const spec = await loadSpec(source);
		const endpoints = extractEndpoints(spec);
		const schemas = extractSchemas(spec);
		onProgress?.({ phase: "parsed", message: `Found ${endpoints.length} endpoints, ${schemas.length} schemas` });

		const endpointDocs = endpoints.map((e) => endpointToDocument(e, apiName));
		const schemaDocs = schemas.map((s) => schemaToDocument(s, apiName));
		const allDocs = [...endpointDocs, ...schemaDocs];

		if (!opts?.skipDelete) {
			onProgress?.({ phase: "deleting", message: "Removing old data..." });
			await this.#store.deleteApi(apiName);
		}

		await this.#store.upsert(allDocs, (done, total, storePhase) => {
			onProgress?.({ phase: storePhase, message: `${storePhase === "embedding" ? "Embedding" : "Storing"} ${done}/${total}`, done, total });
		});

		const summary = { api: apiName, endpointsIngested: endpointDocs.length, schemasIngested: schemaDocs.length, total: allDocs.length };
		onProgress?.({ phase: "done", message: `Done — ${summary.endpointsIngested} endpoints, ${summary.schemasIngested} schemas` });
		return summary;
	}

	async ingestContent(
		raw: string,
		format: "yaml" | "json",
		apiName: string,
		onProgress?: (event: ProgressEvent) => void,
	): Promise<IngestSummary> {
		onProgress?.({ phase: "parsing", message: "Parsing spec..." });
		const spec = await parseSpecContent(raw, format);
		const endpoints = extractEndpoints(spec);
		const schemas = extractSchemas(spec);
		onProgress?.({ phase: "parsed", message: `Found ${endpoints.length} endpoints, ${schemas.length} schemas` });

		const endpointDocs = endpoints.map((e) => endpointToDocument(e, apiName));
		const schemaDocs = schemas.map((s) => schemaToDocument(s, apiName));
		const allDocs = [...endpointDocs, ...schemaDocs];

		onProgress?.({ phase: "deleting", message: "Removing old data..." });
		await this.#store.deleteApi(apiName);

		await this.#store.upsert(allDocs, (done, total, storePhase) => {
			onProgress?.({ phase: storePhase, message: `${storePhase === "embedding" ? "Embedding" : "Storing"} ${done}/${total}`, done, total });
		});

		const summary = { api: apiName, endpointsIngested: endpointDocs.length, schemasIngested: schemaDocs.length, total: allDocs.length };
		onProgress?.({ phase: "done", message: `Done — ${summary.endpointsIngested} endpoints, ${summary.schemasIngested} schemas` });
		return summary;
	}

	// ------------------------------------------------------------------
	// Search
	// ------------------------------------------------------------------

	async searchEndpoints(
		query: string,
		api?: string,
		method?: string,
		tag?: string,
		n: number = 2,
		maxDistance: number = MAX_DISTANCE,
	): Promise<QueryResult[]> {
		const where = buildWhere({ type: "endpoint", api, method });
		let results = await this.#store.query(query, n * 4, where ?? undefined);
		results = results.filter((r) => (r.distance ?? 1) <= maxDistance);
		if (tag) {
			const tagLower = tag.toLowerCase();
			results = results.filter(
				(r) => r.metadata.tags?.toLowerCase().includes(tagLower)
			);
		}
		results = applyHybridBoost(query, results);
		results = penalizeUnstable(results);
		results = deduplicateByPathPrefix(results);
		return results.slice(0, n);
	}

	async searchSchemas(
		query: string,
		api?: string,
		n: number = 2,
		maxDistance: number = MAX_DISTANCE,
	): Promise<QueryResult[]> {
		const where = buildWhere({ type: "schema", api });
		let results = await this.#store.query(query, n * 3, where ?? undefined);
		results = results.filter((r) => (r.distance ?? 1) <= maxDistance);
		return results.slice(0, n);
	}

	async getEndpoint(
		path: string,
		method: string,
		api?: string,
	): Promise<DocumentResult | null> {
		if (api) {
			const docId = `${api}:endpoint:${method.toUpperCase()}:${path}`;
			return this.#store.getById(docId);
		}

		const where = {
			$and: [
				{ type: "endpoint" },
				{ method: method.toUpperCase() },
				{ path },
			],
		};
		const results = await this.#store.query(`${method.toUpperCase()} ${path}`, 1, where);
		return results[0] ?? null;
	}

	// ------------------------------------------------------------------
	// Metadata
	// ------------------------------------------------------------------

	async deleteApi(apiName: string): Promise<void> {
		await this.#store.deleteApi(apiName);
	}

	async listApis(): Promise<ApiInfo[]> {
		return this.#store.listApis();
	}

	async listEndpoints(apiName: string): Promise<DocumentResult[]> {
		const docs = await this.#store.getAll(apiName);
		return docs.filter((d) => d.metadata.type === "endpoint");
	}
}

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

const MAX_DISTANCE = 0.75;

// ---------------------------------------------------------------------------
// Internal helpers
// ---------------------------------------------------------------------------

interface WhereFilters {
	type?: string;
	api?: string;
	method?: string;
}

// Hybrid boost: apply rule-based score adjustments on top of semantic distance.
// Currently handles "create lxc/container" → boost POST /nodes/{node}/lxc exact match.
function applyHybridBoost(query: string, results: QueryResult[]): QueryResult[] {
	const q = query.toLowerCase();
	const isCreateLxc = /\bcreate\b/.test(q) && /\b(lxc|container)\b/.test(q);
	if (!isCreateLxc) return results;

	const boosted = results.map((r) => {
		const method = r.metadata.method ?? "";
		const path = r.metadata.path ?? "";
		// Match exact collection endpoint: POST /nodes/{node}/lxc (no trailing segments)
		const isTarget = method === "POST" && /^\/nodes\/\{[^}]+\}\/lxc$/.test(path);
		if (isTarget) return { ...r, distance: Math.max(0, (r.distance ?? 0) - 0.2) };
		return r;
	});

	return boosted.sort((a, b) => (a.distance ?? 0) - (b.distance ?? 0));
}

// Penalize deprecated/unstable/beta endpoints so stable equivalents rank higher.
const UNSTABLE_PATH_RE = /\/(unstable|beta|experimental|preview|alpha)\//i;
const UNSTABLE_TAG_RE = /\b(unstable|beta|experimental|preview|alpha|deprecated)\b/i;
const UNSTABLE_PENALTY = 0.15;

function penalizeUnstable(results: QueryResult[]): QueryResult[] {
	return results
		.map((r) => {
			const isUnstable =
				r.metadata.deprecated === "true" ||
				UNSTABLE_PATH_RE.test(r.metadata.path ?? "") ||
				UNSTABLE_TAG_RE.test(r.metadata.tags ?? "");
			if (!isUnstable) return r;
			return { ...r, distance: (r.distance ?? 0) + UNSTABLE_PENALTY };
		})
		.sort((a, b) => (a.distance ?? 0) - (b.distance ?? 0));
}

// Deduplicate results to one entry per path prefix, keeping the best score.
// Prefix = path up to (but not including) the second path parameter.
// e.g. /nodes/{node}/lxc/{vmid}/status/start → /nodes/{node}/lxc
function deduplicateByPathPrefix(results: QueryResult[]): QueryResult[] {
	const seen = new Map<string, QueryResult>();
	for (const r of results) {
		const key = pathPrefix(r.metadata.path ?? "");
		const existing = seen.get(key);
		if (!existing || (r.distance ?? 1) < (existing.distance ?? 1)) {
			seen.set(key, r);
		}
	}
	// Preserve original order (best score first)
	return results.filter((r) => seen.get(pathPrefix(r.metadata.path ?? "")) === r);
}

function pathPrefix(path: string): string {
	const segments = path.split("/").filter(Boolean);
	let paramCount = 0;
	const kept: string[] = [];
	for (const seg of segments) {
		if (/^\{[^}]+\}$/.test(seg) && ++paramCount >= 2) break;
		kept.push(seg);
	}
	return "/" + kept.join("/");
}

function buildWhere(filters: WhereFilters): Record<string, unknown> | null {
	const clauses: Record<string, string>[] = [];

	if (filters.type) clauses.push({ type: filters.type });
	if (filters.api) clauses.push({ api: filters.api });
	if (filters.method) clauses.push({ method: filters.method.toUpperCase() });

	if (clauses.length === 0) return null;
	if (clauses.length === 1) return clauses[0];
	return { $and: clauses };
}
